File: 66-plus-one/plus-one.py
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        if all(digit == 9 for digit in digits):
            return [1] + [0] * len(digits)
        
        for i in range(len(digits) - 1, -1, -1):
            digit = digits[i]
            if digit < 9:
                digits[i] += 1
                break
            
            # Carry the '1' over to the next (higher-significant) digit
            digits[i] = 0
        
        return digits

File: 66-plus-one/plus-one.cs
public class Solution {
    public int[] PlusOne(int[] digits) {
        return PlusOneRec(digits, digits.Length - 1);
    }
    private int[] PlusOneRec(int[] digits, int index)
    {
        if (digits[index] < 9)
        {
            digits[index]++;
            return digits;
        }

        // digits[index] == 9
        digits[index] = 0;
        if (index == 0)
        {
            int[] new_digits = new int[digits.Length + 1];
            new_digits[0] = 1;
            for (int i = 0; i < digits.Length; i++)
                new_digits[i + 1] = digits[i];
            
            return new_digits;
        }

        return PlusOneRec(digits, index - 1);
    }
}